Mtc 20200310

landonreed opened the pull request:
### Checklist

- [x] Appropriate branch selected _(all PRs must first be merged to `dev` before they can be merged to `master`)_
- [x] Any modified or new methods or classes have helpful JavaDoc and code is thoroughly commented
- [x] The description lists all applicable issues this PR seeks to resolve
- [x] The description lists any configuration setting(s) that differ from the default settings
- [x] All tests and CI builds passing

### Description

re ibi-group/datatools-ui#544.

landonreed committed:
fix(logback.xml): fix logback formatting for IDE links

landonreed committed:
refactor(MergeFeedsJob): set retrieval method on feed merge; store regional merge in fs

re ibi-group/datatools-ui#544

landonreed committed:
refactor(ProjectController): fix logMessageAndHalt invocation

landonreed committed:
refactor(MergeFeedsJob): add storeNewVersion option

binh-dam-ibigroup commented on src/main/java/com/conveyal/datatools/manager/jobs/MergeFeedsType.java:
```diff
@@ -2,5 +2,5 @@
 
 public enum MergeFeedsType {
```
Add JSDoc for this enum.

binh-dam-ibigroup commented on src/main/java/com/conveyal/datatools/manager/models/Model.java:
```diff
@@ -39,6 +39,8 @@ public Model () {
 
     public String id;
 
+    public String retrieveId() { return this.id; }
```
This should be named `getId()`.

binh-dam-ibigroup replied:
(and `retrieveNotes()` should be renamed `getNotes()`)

landonreed replied:
Getters unfortunately cannot be named `getId` due to issues with the Mongo codecs (same with setters). Perhaps there should be comments on these `retrieve*` methods instead.

landonreed replied:
I could just remove the getter and change the lambda syntax.

binh-dam-ibigroup commented on src/main/java/com/conveyal/datatools/manager/models/Project.java:
```diff
@@ -67,6 +67,12 @@
+     * Feed source in which to store regionally merged GTFS feeds. If specified, during a regional feed merge all feeds
+     * in the project will be merged except for the feed source defined here.
+     */
+    public String regionalFeedSourceId;
```
I am not sure why getters and setters were not set up in this entire class.

binh-dam-ibigroup commented on src/main/java/com/conveyal/datatools/manager/controllers/api/ProjectController.java:
```diff
@@ -219,21 +220,43 @@ private static Project checkProjectPermissions(Request req, Project project, Str
         }
+        // Check that the latest regionally merged feed does not already contain input feed versions.
+        if (project.regionalFeedSourceId != null) {
+            Set<String> versionIds = feedVersions.stream().map(FeedVersion::retrieveId).collect(Collectors.toSet());
```
Change to `FeedVersion::getId` (see comment in Model.java)

binh-dam-ibigroup commented on src/main/java/com/conveyal/datatools/manager/models/FeedVersion.java:
```diff
@@ -50,6 +51,10 @@
+    /**
+     * Input feed versions used to create a merged version.
+     */
+    public Set<String> inputVersions;
```
Replace public field with getter and setter.

binh-dam-ibigroup reviewed (changes requested):
Fix some method names, add enum JSDoc.
Should public fields be turned into getters/setters? (should still work with 
jackson)

evansiroky commented on src/main/resources/logback.xml:
```diff
@@ -1,9 +1,9 @@
-            <Pattern>%d{MMM dd HH:mm:ss.SSS} [%thread] %-5level %logger{36}:%L - %msg%n</Pattern>
-        </layout>
+        <encoder class="ch.qos.logback.classic.encoder.PatternLayoutEncoder">
+            <pattern>%highlight(%-5level) %d{yyyy-MM-dd'T'HH:mm:ss.SSS} %yellow([%thread]) %blue(%logger{36}\(%file:%line\)) %msg%n</pattern>
```
+1 for this date format

evansiroky commented on src/main/java/com/conveyal/datatools/manager/jobs/MergeFeedsJob.java:
```diff
@@ -132,29 +135,57 @@
+        this.feedSource = mergeType.equals(REGIONAL)
+            ? regionalFeedSource
+            : feedVersions.iterator().next().parentFeedSource();
+        // Merged version will be null if the new version should not be stored.
```
Can you also write some JavaDoc about how this class variable can be null up in the variable definition?

binh-dam-ibigroup committed:
fix(FeedVersion, gtfsplus.yml): Add call to MTCValidator, change gtfsplus.yml to match MTC guideline

binh-dam-ibigroup committed:
style(gtfsplus.yml): Order file top-level entries alphabetically.

binh-dam-ibigroup committed:
fix(FeedVersion): Run MTCValidator only if 'mtc' extension is enabled.

binh-dam-ibigroup committed:
build(maven): Update gtfs-lib version

landonreed committed:
Merge pull request #297 from ibi-group/mtc-config

Add call to MTCValidator

landonreed committed:
build(deps): bump gtfs-lib to 6.0.1 for empty line validation

re conveyal/gtfs-lib#280

landonreed committed:
build(pom): update comment

this throwaway commit is basically being used to trigger a travis build that didn't start due to a
GitHub outage earlier today

landonreed committed:
Merge branch 'fix-rider-cat' into mtc-20200310

landonreed committed:
Merge branch 'dev' into mtc-20200310

landonreed committed:
feat(gtfs-transform): add arbitrary transforms for imported versions

This adds the ability to define a list of arbitrary transformations that should be applied to each
new feed version loaded for a particular feed source. The primary use case for now is limited to
adding/replacing a file (in this case a GTFS+) file into each new version.

re ibi-group/datatools-ui#544

landonreed committed:
refactor(gtfs-transform): refactor FeedTransformation into abstract class

landonreed committed:
refactor(gtfs-transform): change how things are structured a bit

landonreed committed:
fix(deps): bump gtfs-lib to 6.0.3

fix #313

landonreed committed:
refactor(gtfs-transform): refactor for Jackson polymorphism handling

landonreed committed:
Merge branch 'dev' into gtfs-transform

landonreed committed:
refactor: remove unused imports

landonreed committed:
refactor(gtfs-transform): address PR #309 comments

landonreed committed:
test: fix tests, add trip to make test more reliable

landonreed committed:
refactor(gtfs-transform): add version_clone test, ReplaceFileTransformation -> ReplaceFileFromVersio

landonreed committed:
refactor(FeedTransformation): rename subtype name for jackson

landonreed committed:
ci(travis): log git describe at top of build

landonreed committed:
ci(travis): remove git describe, which does not match deployed jar

landonreed committed:
refactor(gtfs-transform): remove SQL injection risks

landonreed committed:
refactor: fix field name

landonreed committed:
refactor(gtfs-transform): preserve results of feed transformations

landonreed committed:
refactor: address PR comments

landonreed committed:
refactor: address PR #309 comments

landonreed committed:
Merge pull request #309 from ibi-group/gtfs-transform

GTFS transform

landonreed committed:
Merge branch 'dev' into mtc-20200310

landonreed committed:
refactor: address PR #288 comments
